### PR TITLE
Add poseidon-deployer

### PR DESCRIPTION
**Pull request by baumstern**
poseidon-deployer is a set of Hardhat tasks for deploying [Poseidon](https://www.poseidon-hash.info) hash contract(s). It uses [circomlibjs](https://github.com/iden3/circomlibjs) to generate bytecode of Poseidon contract.

- MACI contracts depend on Poseidon contracts
- This will help automate that process, reduce gas costs from re-deploying contracts & reduce errors

### CODE REVIEW

File: devops/poseidon-deployer/contracts/overkill.sol [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    /// -----------------------------------------------------------------------
+    /// Library Usage
+    /// -----------------------------------------------------------------------
+
+    using LibClone for address; //hacks
+
+    /// -----------------------------------------------------------------------
+    /// Events
+    /// -----------------------------------------------------------------------
+
```
Comment (daodesigner) [inline devops/poseidon-deployer/contracts/overkill.sol:22]: 🧐 think we missed T1

File: devops/poseidon-deployer/tasks/deployPoseidon.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    default:
+      poseidonContracts = await hre.run("deploy-poseidon:multipleTx");
+  }
+
+  return poseidonContracts;
+});
+
+
+subtask("deploy-poseidon:multipleTx")
```
Comment (daodesigner) [inline devops/poseidon-deployer/tasks/deployPoseidon.ts:35]: deploy-poseidon:EOA

File: devops/poseidon-deployer/tasks/deployPoseidon.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import { task, subtask } from "hardhat/config";
+import { Contract } from "ethers";
+
+import { verifyPoseidonContracts } from "./utils";
+
+
+task("deploy-poseidon", "Deploy poseidon contracts by sending multiple txs or calling multicall")
+.addFlag("multicall", "Use this flag to invoke multicall")
```
Comment (daodesigner) [inline devops/poseidon-deployer/tasks/deployPoseidon.ts:8]: Remove this flag add a flag for each PoseidonT[x] supported

File: devops/poseidon-deployer/tasks/deployPoseidon.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    // deploy contracts
+  const results = await poseidonFactory.multicall(calls);
+  console.log("multicall result: " + JSON.stringify(results));
+  } catch(e){
+    throw new Error('failed to process multicall. Contracts had not been deployed. Maybe salt has been already used to deploy poseidon: \n' + e);
+  }
+  
+  const addresses = Array.from(poseidonAddressBook.values());
+
+  // Ensure deployed contracts are correctly process poseidon functions
```
Comment (daodesigner) [inline devops/poseidon-deployer/tasks/deployPoseidon.ts:89]: Add what the verification being run is (assuming it's hashing all zeros)

File: devops/poseidon-deployer/tasks/utils.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  const poseidonReference = await buildPoseidonOpt();
+  let poseidonContracts = new Map();
+  let input = [];
+
+  for (let TN = 2; TN <= 7; TN++) {
+    const address =addresses[`${TN-2}`];
+    poseidonContracts[`T${TN}`] = await ethers.getContractAt(`PoseidonT${TN}`, address);
+    input.push(`${TN}`);
+   
+    const expectedHash = poseidonReference.F.toString(poseidonReference(input));
```
Comment (daodesigner) [inline devops/poseidon-deployer/tasks/utils.ts:17]: NIT: Add what the expect hashes/inputs actually are for each in comments

File: devops/poseidon-deployer/test/poseidon.test.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  const contents = 
+      `[poseidonAddresses]
+      T2 = "${poseidonAddressBook.get("T2")}"
+      T3 = "${poseidonAddressBook.get("T3")}"
+      T4 = "${poseidonAddressBook.get("T4")}"
+      T5 = "${poseidonAddressBook.get("T5")}"
+      T6 = "${poseidonAddressBook.get("T6")}"
+      T7 = "${poseidonAddressBook.get("T7")}"`
+      
+      writeFileSync(poseidonAddressBookFilePath, contents);
```
Comment (daodesigner) [inline devops/poseidon-deployer/test/poseidon.test.ts:36]: Run linter

File: devops/poseidon-deployer/contracts/overkill.sol (hunk 1 above)
Comment (gurrpi) [reply devops/poseidon-deployer/contracts/overkill.sol]: T1 makes number of input to 0 (i.e. no input) due to [`const t = nInputs + 1;`](https://github.com/iden3/circomlibjs/blob/main/src/poseidon_gencontract.js#L26)

File: devops/poseidon-deployer/tasks/deployPoseidon.ts (hunk 2 above)
Comment (gurrpi) [reply devops/poseidon-deployer/tasks/deployPoseidon.ts]: Reflected:
 https://github.com/privacy-scaling-explorations/maci/blob/3f753a1792931a5910c3614db2426bef2bb134c1/devops/poseidon-deployer/tasks/deployPoseidon.ts#L5

File: devops/poseidon-deployer/tasks/deployPoseidon.ts (hunk 3 above)
Comment (gurrpi) [reply devops/poseidon-deployer/tasks/deployPoseidon.ts]: Created `deploy-poseidon:callPoseidonFactory` command to deploy each PoseidonT[x] by multicall:

https://github.com/privacy-scaling-explorations/maci/blob/3f753a1792931a5910c3614db2426bef2bb134c1/devops/poseidon-deployer/tasks/deployPoseidon.ts#L29

File: devops/poseidon-deployer/tasks/deployPoseidon.ts (hunk 4 above)
Comment (gurrpi) [reply devops/poseidon-deployer/tasks/deployPoseidon.ts]: Mentioned verification is done by comparing to reference implementation:

https://github.com/privacy-scaling-explorations/maci/blob/3f753a1792931a5910c3614db2426bef2bb134c1/devops/poseidon-deployer/tasks/deployPoseidon.ts#L100-L106

File: devops/poseidon-deployer/tasks/utils.ts (hunk 5 above)
Comment (gurrpi) [reply devops/poseidon-deployer/tasks/utils.ts:49]: Reflected:

https://github.com/privacy-scaling-explorations/maci/blob/3f753a1792931a5910c3614db2426bef2bb134c1/devops/poseidon-deployer/tasks/utils.ts#L40-L53

File: devops/poseidon-deployer/test/poseidon.test.ts (hunk 6 above)
Comment (gurrpi) [reply devops/poseidon-deployer/test/poseidon.test.ts]: Had run prettier and ESLint in 3f753a1792931a5910c3614db2426bef2bb134c1

**Comment by baumstern** [comment]
Reopening this PR as it appears the expected refactor by @ctrlc03 hasn't materialized, and there seems to be no redundancy as of now. This implementation is feature-complete, comes with comprehensive documentation, and has been rigorously tested. Would appreciate a review @ctrlc03

**Comment by samajammin** [comment]
@baumstern I'm not clear on what problem this solves. Is there an associated issue for this PR? Or could you please explain the added benefit here? Pardon my ignorance 😄 just looking to understand the benefit of this.

**Comment by baumstern** [comment]
> @baumstern I'm not clear on what problem this solves. Is there an associated issue for this PR? Or could you please explain the added benefit here? Pardon my ignorance 😄 just looking to understand the benefit of this.

As @ctrlc03 explained in our sprint planning meeting today, the high-level idea is to save approximately ~98%(reducing gas costs from 18,761,539 to 352,286)  on deployment gas costs for crypto contracts by providing reusable proxy contracts. This approach also saves developers hours of trial and error typically needed to deploy Poseidon contracts correctly.

**Comment by baumstern** [comment]
Closing this PR since it has ported into https://github.com/baumstern/poseidon-deployer

**Pull request closed without merging**